src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+    kz,
+    // wave number (different from kz only if we are taking a part
+    // of the z-domain [and not from 0 to 2*pi])
+    kz * kwaveFactor, global_flags, inner_boundary_flags,
+    outer_boundary_flags, &A, &C, &D);
+
+    // Patch up internal boundaries
+    if(not localmesh->lastX()) { 
+      for(int ix = localmesh->xend+1; ix<localmesh->LocalNx ; ix++) {
+	avec(kz,ix) = 0;
+	bvec(kz,ix) = 1;
+	cvec(kz,ix) = 0;
+	bcmplx(kz,ix) = 0;
+      }
+    } 
+    if(not localmesh->firstX()) { 
+      for(int ix = 0; ix<localmesh->xstart ; ix++) {
+	avec(kz,ix) = 0;
+	bvec(kz,ix) = 1;
+	cvec(kz,ix) = 0;
+	bcmplx(kz,ix) = 0;
+      }
+    }
+  }
+  ///SCOREP_USER_REGION_END(kzinit);
+  ///SCOREP_USER_REGION_DEFINE(initlevels);
+  ///SCOREP_USER_REGION_BEGIN(initlevels, "init levels",///SCOREP_USER_REGION_TYPE_COMMON);
+
+  // Initialize levels. Note that the finest grid (level 0) has a different
+  // routine to coarse grids (which generally depend on the grid one step
+  // finer than itself).
+  //
+  // If the operator to invert doesn't change from one timestep to another,
+  // much of the information for each level may be stored. Data that cannot
+  // be cached (e.g. the changing right-hand sides) is calculated in init_rhs
+  // below.
+  if(first_call(jy,0) || not store_coefficients ){
+
+    init(levels[0], ncx, jy, avec, bvec, cvec, xs, xe);
+
+    int ncx_coarse = ncx; //(xe-xs+1)/2 + xs + ncx - xe - 1;
+    if(max_level>0){
+      for(int l = 1; l<=max_level; l++){
+	ncx_coarse = (ncx_coarse-4)/2+4;
+	init(levels[l], levels[l-1], ncx_coarse, xs, ncx_coarse-3,l,jy); //FIXME assumes mgy=2
+      }
+    }
+  }
+
+  // Compute coefficients that depend on the right-hand side and which
+  // therefore change every time.
+  init_rhs(levels[0], jy, bcmplx);
+
+  ///SCOREP_USER_REGION_END(initlevels);
+
+  ///SCOREP_USER_REGION_DEFINE(setsoln);
+  ///SCOREP_USER_REGION_BEGIN(setsoln, "set level 0 solution",///SCOREP_USER_REGION_TYPE_COMMON);
+  // Set initial values with cached values
+  for(int ix=0; ix<4; ix++){
+    for(int kz=0; kz<nmode; kz++){
+      levels[0].xloc(ix,kz) = x0saved(jy, ix, kz);
+    }
+  }
+  ///SCOREP_USER_REGION_END(setsoln);
+  ///SCOREP_USER_REGION_DEFINE(initwhileloop);
+  ///SCOREP_USER_REGION_BEGIN(initwhileloop, "init while loop",///SCOREP_USER_REGION_TYPE_COMMON);
+
+  int count = 0; // Total iteration count
+  int subcount = 0; // Count of iterations on a level
+  int cyclecount = 0; // Number of multigrid cycles
+  int cycle_eta = 0; // Predicted finishing cycle
+  int current_level = 0;
+  bool down = true;
+  auto converged = Array<bool>(nmode);
+
+  auto error_abs = Array<BoutReal>(nmode);
+  auto error_abs_old = Array<BoutReal>(nmode);
+  auto error_rel = Array<BoutReal>(nmode);
+  auto error_rel_old = Array<BoutReal>(nmode);
+  for(int kz=0; kz<nmode; kz++){
+    converged[kz] = false;
+    error_abs[kz] = 1e20;
+    error_abs_old[kz] = 1e20;
+    error_rel[kz] = 1e20;
+    error_rel_old[kz] = 1e20;
+  }
```

Comment:
This works for the errors, but segfaults for the `Array<bool> converged`.

Reply:
Segfaults on the assignment? Or somewhere else?

Reply:
Oops, I see the problem now. `Array` only has a copy-assignment operator, so

```cpp
Array<bool> thing(12);
thing = true;
```
actually becomes

```cpp
Array<bool> thing(12);
thing = Array<bool>(1);
```

because `Array(size_type len)` is not explicit and `bool` is implicitly convertible to `size_type` :(

Maybe we need a method like [`std::vector::assign`](https://en.cppreference.com/w/cpp/container/vector/assign)